[WIP]Add a new helper method for evalf

sidhu1012 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #20479 

#### Brief description of what is fixed or changed
Added a new helper method for `evalf` to make `evalf` consistent without compromising `precision` and `options` parameter.

#### Other comments
Read discussion #20479 

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* core
    * Added a new helper method for evalf()
<!-- END RELEASE NOTES -->

sidhu1012 committed:
Add a new helper method for evalf

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -196,8 +196,17 @@ def negated(self):
+        #return self.func(*[evalf(s, prec, options) for s in self.args])
+        return self.func(*[s.evalf(n, subs=subs, maxn=options['maxn'], chop=options['chop'],
+                            strict=options['strict'], quad=quad,
+                            verbose=options['verbose']) for s in self.args])
```
This is a horrible way to recurse. We should make it so that `_eval_evalf_options` takes the same arguments as the `evalf` function and then recurse using the `evalf` function.

sidhu1012 replied:
Yes, agreed.

czgdp1807 commented:
Do we need a release notes entry for this?

sidhu1012 committed:
Break .evalf in 2 parts to control the pre and post processing

sidhu1012 commented:
@oscarbenjamin I broke `.evalf` function in 2 parts, one part is the public `.evalf` function which handles the preprocessing and calls the other part. The other part calls the `evalf` from `sympy.core.evalf` and handles some corner cases which are required in the codebase.

oscarbenjamin commented:
This looks like the right idea but there needs to be comments and docs explaining how to use this. We also need to make sure that this works with all the different combinations of how downstream libraries and users might be depending on the existing behaviour. There hasn't been clear documentation for users so there can be many combinations of how they might have implemented these things (there is inconsistency in the sympy codebase so I expect inconsistency elsewhere as well).

Somewhere the documentation should explain how Basic subclasses should implement their own evalf routines probably in the class docstring for `EvalfMixin` which is currently very short:
https://docs.sympy.org/latest/modules/core.html#sympy.core.evalf.EvalfMixin

I think a docs example implementing a simple `Basic` subclass with `_eval_evalf_options` would be useful. This should clarify that a subclass should inherit `EvalfMixin` and should override the `_eval_evalf_options` method and should recurse using `._evalf_options`. There should also be notes explaining that the `_eval_evalf` and `_evalf` methods are retained/used only for backwards compatibility.

There should also be some reference to `EvalfMixin` (perhaps a see also) in the `evalf` docstring referring to there for how to implement evalf behaviour in user-created classes.

Anywhere in the code where `_eval_evalf_options` or `_eval` or `_eval_evalf` is used should have comments discussing why the different methods are being called (e.g. for backwards compatibility). Each of these methods should have a docstring indicating what it is to be used for and whether it is recommended or deprecated.

To ensure backwards compatibility we need to test the behaviour of different possible subclasses that users might have made in the past. A user might have made a class which is

1. An Expr subclass
2. A Basic subclass
3. A Basic subclass that also subclasses EvalfMixin

That subclass might implement:

1. `._eval_evalf`
2. `._eval`
3. `.evalf`
4. Some combination of the above
4. None of the above

That gives a lot of combinations for how someone might have implemented this. We should test that all combinations still give the same results after this change. We should test the expected behaviour for precision, subs etc.

We also need to test in reverse. Some downstream user code might be calling methods like `_eval_evalf` or `_eval` on standard sympy objects both Basic and Expr. We need to test that their code will still work correctly (or at least the same as before) when given a class that implements `_eval_evalf_options` instead.

Also a more extensive release note is needed. We need to make sure that anyone who has implemented an `_eval_evalf` method knows that they should add an `_eval_evalf_options` method as well. They might need to add both methods in order to retain compatibility with different sympy versions (so we need to test that we get the expected behaviour when *both* methods are defined). Also if they have used `.evalf` or `._evalf` or `._eval_evalf` to recurse then they also need to change their code and there needs to be an explanation of how to change it while maintaining compatibility with older sympy versions (and we need to test that their changed code works with the new sympy code).

We should clearly indicate that the use of either `_eval_evalf` or `_evalf` is deprecated although I don't think that `DeprecationWarning` should be used. Instead the docs can use the `.. deprecated::` directive and the code should have comments clearly indicating where deprecated code is being called for backwards compatibility.

I think it probably makes sense to have a function `evalf_options` in the `evalf` module whose purpose is to check for the different methods `._eval_evalf_options`, `._eval_evalf`, '._eval` and call the appropriate implementation. Then we can use that in the evalf module any time we need to call one of these methods. Potentially that is also helpful for users who want to support the new sympy code as well as the old because we can suggest for them to do something like (not sure if this actually works in all cases):
```python
try:
    # sympy version >= 1.8:
    from sympy.core.evalf import evalf_options
except ImportError:
    # sympy version < 1.8
    def evalf_options(obj, prec, options):
        return obj._eval_evalf(prec)

class MyClass(Expr):
    def _eval_evalf_options(self, prec, options):
        args = [evalf_options(a, prec, options) for a in self.args]
        # new code that makes use of options

    def _eval_evalf(self, prec):
        # old version that doesn't make use of options
```
Basically it is not enough to ensure that the sympy codebase is internally consistent because I think this will be used a lot by downstream code. We need to make sure that things that used to work continue to work. We also need to document for users how they should ideally implement things and how to make their code work with both old and new version of sympy. We also need tests for everything.

sidhu1012 commented:
@oscarbenjamin I will update the docstring wherever necessary.


> That gives a lot of combinations for how someone might have implemented this. We should test that all combinations still give the same results after this change. We should test the expected behaviour for precision, subs etc.
> 
> We also need to test in reverse. Some downstream user code might be calling methods like `_eval_evalf` or `_eval` on standard sympy objects both Basic and Expr. We need to test that their code will still work correctly (or at least the same as before) when given a class that implements `_eval_evalf_options` instead.
> 

`_evalf_options` is a private function which is called by public `.evalf` so that doesn't need and changes downstream , the only change required is to replace `_eval_evalf` with `eval_evalf_options` so that `options` parameters are considered and recurse on `evalf_options` to skip preprocessing.
The changes that I have done would first look for `_eval_evalf_options` and if it's not defined(gives `None` or class is using `_eval_evalf`) then it would revert back to `eval_evalf`, thus won't break backward compatibility but would just prefer `_eval_evalf_options` over `_eval_evalf` . `_eval` just calls  `_eval_evalf` so that won't be an isse.

So I think all possible combinations of subclasses using `evalf` have been covered.



> I think it probably makes sense to have a function `evalf_options` in the `evalf` module whose purpose is to check for the different methods `._eval_evalf_options`, `._eval_evalf`, '._eval` and call the appropriate implementation. Then we can use that in the evalf module any time we need to call one of these methods. Potentially that is also helpful for users who want to support the new sympy code as well as the old because we can suggest for them to do something like (not sure if this actually works in all cases):

I don't think it's a good idea because `._evalf_options` and `evalf()` would take care of implemented methods and would call them appropriately .

sidhu1012 committed:
Improved docstring and comments

sidhu1012 committed:
maybe fix indentation error in docstring

sidhu1012 commented:
@oscarbenjamin How to define classes in docstring?

oscarbenjamin commented:
> So I think all possible combinations of subclasses using `evalf` have been covered.

Can you add tests for all cases?

sidhu1012 commented:
> Can you add tests for all cases?

Yes I would in next commits.

sidhu1012 committed:
Fixed example

sidhu1012 committed:
Added test cases to check all posible combinations of EvalfMixin functions for backwards compatibility

change deprection

sidhu1012 commented:
@oscarbenjamin I have added test cases for all possible combinations of EvalfMixin. Please review.

sidhu1012 commented:
I replaced `eval_evalf` with `eval_evalf_options` randomly at many places and is working fine and isn't breaking code plus it also corrected precision at many places and all `options` are working fine. Example `subs` with `FiniteSet`.

sidhu1012 commented:
@oscarbenjamin Ready for review!

oscarbenjamin commented:
These tests are quite repetitive. You should combine them into a single function and make a superclass to absorb the repeated code.

You are testing lots of combinations of classes that all only inherit from `EvalfMixin`. That's one possible case but there are others such as a subclass of Expr (which already subclasses EvalfMixin) and a subclass of Basic that does subclass EvalfMixin and also one that does not.

Crucially we need to think about cases where someone has not subclassed EvalfMixin at all and has just implemented a method like `_eval_evalf`. In current master I can do this:
```
In [70]: class A(Basic): 
    ...:     def _evalf(self, prec): 
    ...:         return Float('2.0', prec) 
    ...:                                                                                                                          

In [71]: FiniteSet(A()).evalf()                                                                                                   
Out[71]: {2.0}
```
With this PR that gives:
```
AttributeError: 'A' object has no attribute 'evalf'
```

sidhu1012 commented:
> Crucially we need to think about cases where someone has not subclassed EvalfMixin at all and has just implemented a method like `_eval_evalf`. In current master I can do this:
> 
> ```
> In [70]: class A(Basic): 
>     ...:     def _evalf(self, prec): 
>     ...:         return Float('2.0', prec) 
>     ...:                                                                                                                          
> 
> In [71]: FiniteSet(A()).evalf()                                                                                                   
> Out[71]: {2.0}
> ```
> 
> With this PR that gives:
> 
> ```
> AttributeError: 'A' object has no attribute 'evalf'
> ```

In current master it gives `AttributeError: 'A' object has no attribute 'evalf'` and so does in this PR.

sidhu1012 commented:
> > Crucially we need to think about cases where someone has not subclassed EvalfMixin at all and has just implemented a method like `_eval_evalf`. In current master I can do this:
> > ```
> > In [70]: class A(Basic): 
> >     ...:     def _evalf(self, prec): 
> >     ...:         return Float('2.0', prec) 
> >     ...:                                                                                                                          
> > 
> > In [71]: FiniteSet(A()).evalf()                                                                                                   
> > Out[71]: {2.0}
> > ```
> > 
> > 
> > With this PR that gives:
> > ```
> > AttributeError: 'A' object has no attribute 'evalf'
> > ```
> 

In current master it gives `AttributeError: 'A' object has no attribute 'evalf'` and so does in this PR.

sidhu1012 commented:
@oscarbenjamin Since the code you showed is giving the same `AttributeError` should I make it work in this PR ?

sidhu1012 commented:
@oscarbenjamin Since both the master branch and this PR both are giving `AttributeError`  what shall I do?

sidhu1012 committed:
Made _evalf_options out of class for backward compatibility

fixed docs

oscarbenjamin commented:
> In current master it gives `AttributeError: 'A' object has no attribute 'evalf'` and so does in this PR.

Sorry, I should have said that in sympy 1.7 it works. It seems to have already broken on master from another evalf PR.

Maybe that particular case isn't so important but we do need to be careful to make sure that anything that worked in sympy 1.6 or 1.7 still works if possible.

sidhu1012 commented:
@oscarbenjamin I have added some new tests for remaining cases.Please review them, I think it isn't breaking any code and is playing nice.

sidhu1012 committed:
test_typo

sidhu1012 commented:
@oscarbenjamin If these changes are correct then shall I reflect them to entire codebase?

oscarbenjamin commented on sympy/core/evalf.py:
```diff
@@ -1375,9 +1388,77 @@ def evalf(x, prec, options):
         check_target(x, r, prec)
     return r
 
+def _evalf_options(x, prec, options):
```
This function should just be called `evalf_options` i.e. no underscore because it's not private.

oscarbenjamin commented:
The tests for all the different combinations are long and repetitive. It should be possible to combine them together and make them more systematic e.g.:
```
class A: ...
class B: ...
class C: ...

all_classes = [A, B, C]

def test_in_finiteset(cls): ...
def test_in_matrix(cls): ...
def test_in_sin(cls): ...

all_test_funcs = [test_in_finiteset, ...]

for cls in all_classes:
    for func in all_test_funcs:
        func(cls)
```

The most important cases to try and test for are the ones that users are likely to have used and that we want to protect for backwards compatibility. A reasonable guess at that can be looking at the ways that `EvalfMixin` is used around the codebase.

I'm still not really happy with "_eval_evalf_options" as a name but I haven't thought of anything better.

sidhu1012 commented:
> I'm still not really happy with "_eval_evalf_options" as a name but I haven't thought of anything better.


We can do `function overloading` and keep the name `_eval_evalf`.

It would be easier for downstream libraries too because they would need to change just the paramter and recursive call.

oscarbenjamin commented:
> We can do `function overloading` and keep the name `_eval_evalf`.

If downstream code has defined an `_eval_evalf` method that only has a single parameter then calling it with extra arguments will give a TypeError. That means anyone who has made an `_eval_evalf` function would find that their code is broken by the change in sympy.

sidhu1012 committed:
make evalf_options public, added test cases, handled exception and added
back `eval_evalf` to FiniSet for backward compatibility.

sidhu1012 commented:
@oscarbenjamin I have added the test cases , please review.

While adding test cases I realized that we should add `_eval_evalf_options` where ever necessary but shouldn't replace `_eval_evalf` yet for backward compatibility . So instead of replacing `_eval_evalf` , classes should have both the functions.

oscarbenjamin commented on sympy/sets/sets.py:
```diff
@@ -1883,6 +1883,10 @@ def as_relational(self, symbol):
     def compare(self, other):
         return (hash(self) - hash(other))
 
+    def _eval_evalf_options(self, prec, options):
+        from sympy.core.evalf import evalf_options
+        return FiniteSet(*[evalf_options(elem, prec, options) for elem in self])
+
     def _eval_evalf(self, prec):
         return FiniteSet(*[elem.evalf(n=prec_to_dps(prec)) for elem in self])
```
One of these methods should be calling the other.

It probably makes sense to add a default implementation of `_eval_evalf` to `EvalfMixin` that calls `_eval_evalf_options`.

sidhu1012 replied:
We can change implementation of `_eval_evalf` in Sympy codebase but changing default `_eval_evalf` won't help because they are overridden in every class.


We can't even call `_eval_evalf_options`  from `_eval_evalf` because we won't have `options` parameter.

oscarbenjamin replied:
It's helpful to add in the default implementation because it means that if downstream code is calling `_eval_evalf` directly on a sympy class then that will still work. It also means that we don't need to duplicate the methods like above.

sidhu1012 replied:
Current changes in `.evalf()` would call `_eval_evalf_options` first , it fails then would try with `_eval_evalf` making both work.

I added back `_eval_evalf` because if some method class isn't updated accordingly it doesn't fail, to maintain backward compatibility .

We can't call `_eval_evalf_options` in any way from `_eval_evalf` because of missing parameter `options` and would give `TypeError`, whole purpose of `_eval_evalf_options` is to handle `options` because `_eval_evalf` can not.

oscarbenjamin replied:
The default `_eval_evalf` method in `EvalfMixin` could be made to call `_eval_evalf_options` passing the default values for the options.

sidhu1012 replied:
If we change the default `_eval_evalf` method in `EvalfMixin` still we would need to duplicate the method because every every class calls `evalf` on its args differently in `_eval_evalf` that's why they need to be overridden in every class.

as in `FiniteSet` it is -
https://github.com/sympy/sympy/blob/d053c1d4c3ca5420851dcd9a30dfc0e7cee98a48/sympy/sets/sets.py#L1908

but in it `Interval` -
https://github.com/sympy/sympy/blob/d053c1d4c3ca5420851dcd9a30dfc0e7cee98a48/sympy/sets/sets.py#L1106

oscarbenjamin replied:
As long as EvalfMixin makes sure that both of those end up calling `_eval_evalf_options` then there should be no need to duplicate anything.

sidhu1012 replied:
Yes, got it. Sorry mistook `_eval_evalf_options` with `evalf_options`.

sidhu1012 committed:
make `_eval_evalf` call `_eval_evalf_options` by default

sidhu1012 commented:
@oscarbenjamin I think all possible scenarios have been handled so shall I reflect changes everywhere?

oscarbenjamin commented:
I'm not sure I understand the point of the different classes that are tested A, B and C. They seem unnecessarily complicated for what is tested. The most important case to test is missing which is a subclass of Expr that only implements `_eval_evalf`. That can be:
```
class ExprEvalf(Expr):
    def _eval_evalf(self, prec):
        return Rational(1, 3).evalf(prec)
```
You should add to these tests for any class whose `_eval_evalf` methods are changed (`FiniteSet` etc) especially if they change how they recurse.

More comments/notes are needed to explain why some methods are deprecated and why some call others etc. Especially since this is a change compared to before there needs to be a clear explanation of why it is changed.

Make sure to build the html docs and check to see what it looks like there: it is clear to users what they should do and what is deprecated and why etc? Separately you should make sure with comments in the code that it is clear to anyone reading the code why a deprecated method is being called.

There should be three examples in the `EvalfMixin` docstring: a Basic subclass, an Expr subclass and a non-Basic class. Each of them should illustrate the precise minimum that is needed to make evalf work for a class.

oscarbenjamin commented:
Is it necessary to defined `is_number` or `is_Number`? One is used in the docstring and another in the tests.

oscarbenjamin commented on sympy/core/tests/test_evalf.py:
```diff
@@ -579,3 +580,102 @@ def test_issue_13425():
+        pass
+
+    class C(Basic, X):
+        def could_extract_minus_sign(self):
+            return False
+
+        def as_coefficient(self, value):
+            return
```
Why are these methods needed? Are they used by `EvalfMixin`?

sidhu1012 replied:
They were needed in case of sub class of `Basic` in case of `sin`.

oscarbenjamin commented:
> There should be three examples in the `EvalfMixin` docstring: a Basic subclass, an Expr subclass and a non-Basic class. Each of them should illustrate the precise minimum that is needed to make evalf work for a class.

A fourth example is also needed which shows how to write an Expr subclass that works both with old and new sympy versions.

It's important to be clear when making a change like this what has changed and why and how others should manage the change in their own code.

sidhu1012 commented:
> Is it necessary to defined `is_number` or `is_Number`? One is used in the docstring and another in the tests.

Yes they are needed for independent classes inheriting `EvalfMixin`.  `is_number` or `is_Number` and `func(self)` is required.

sidhu1012 committed:
Update docstring with more notes and examples

sidhu1012 commented:
> A fourth example is also needed which shows how to write an Expr subclass that works both with old and new sympy versions.

I'm not sure how to explain this with an example, can you please suggest an example.

sidhu1012 commented:
> I'm not sure I understand the point of the different classes that are tested A, B and C. They seem unnecessarily complicated for what is tested.

Those tests are just checking that `_eval_evalf_options` is working fine for various possible classes.

sidhu1012 committed:
Check `_eval_evalf` works fine for Basic and Expr subclasses respectively

oscarbenjamin commented:
> > A fourth example is also needed which shows how to write an Expr subclass that works both with old and new sympy versions.
> 
> I'm not sure how to explain this with an example, can you please suggest an example.

If a user updates their `_eval_evalf` method to `_eval_evalf_options` then their code will work with the changed version of sympy. It will still be possible to call `_eval_evalf` in the changed version of sympy because `EvalfMixin` provides that. However if they also want to provide `_eval_evalf` to work with older sympy versions then they will need to copy the `_eval_evalf` implementation from here. They will also need a replacement for `evalf_options`. We need to think about how to do this to make it as easy as possible for them and then document that.

sidhu1012 commented:
> If a user updates their `_eval_evalf` method to `_eval_evalf_options` then their code will work with the changed version of sympy. It will still be possible to call `_eval_evalf` in the changed version of sympy because `EvalfMixin` provides that. However if they also want to provide `_eval_evalf` to work with older sympy versions then they will need to copy the `_eval_evalf` implementation from here. They will also need a replacement for `evalf_options`. We need to think about how to do this to make it as easy as possible for them and then document that.


The best way for `.evalf` work with older sympy versions is to implement `_eval_evalf` and recurse using `.evalf()`.
I'll update the docstring!

sidhu1012 committed:
Update docstring

sidhu1012 commented:
> Make sure to build the html docs and check to see what it looks like there: it is clear to users what they should do and what is deprecated and why etc? 

Where should I add html doc and what should be it's content?

sidhu1012 commented:
@oscarbenjamin I have updated the test cases as well as docstrings. Please go through them once and review them.

sidhu1012 commented:
> Make sure to build the html docs and check to see what it looks like there: it is clear to users what they should do and what is deprecated and why etc?

@oscarbenjamin Can you please help me out with it. I'm not sure what should I do for this.

sidhu1012 committed:
Extend `eval_evalf_options` - I

sidhu1012 commented:
@czgdp1807 A couple of stochastic process tests are failing I am not sure why as I didn't do any stats related change. Can you have a look once. I'm trying on my end.

czgdp1807 commented:
```python
executable:         /usr/bin/python3  (3.6.9-final-0) [CPython]
architecture:       64-bit
cache:              yes
ground types:       python 
numpy:              1.19.2
hash randomization: on (PYTHONHASHSEED=790239741)

sympy/stats/stochastic_process_types.py[11] ...........                      [OK]
```
On latest `master` the doctests are passing. So probably, the changes are affecting the results generated by other modules.

oscarbenjamin commented:
The tests that are run on github are not run directly on the code here. Rather a branch is created from master and this PR is merged into that branch and then the tests are run on that.

If you are seeing failures that you can't reproduce locally you might need to merge in master to see them.

sidhu1012 commented:
> > Make sure to build the html docs and check to see what it looks like there: it is clear to users what they should do and what is deprecated and why etc? 
> 
> Where should I add html doc and what should be it's content?
> 
> 

@oscarbenjamin Can you please help me out with it, as I'm not sure how to do this.

oscarbenjamin commented:
The html docs are built form the rst files in the `doc` folder. They also include the content of the docstrings. I'm suggesting that you should build the docs, open the html in a browser and read it to see if someone looking at the docs would understand how to implement `.evalf` correctly.

sidhu1012 committed:
Revert "Extend `eval_evalf_options` - I"

This reverts commit f8d62015dc46df3b1e893fcc2a361a44b6558207.

sidhu1012 commented:
@oscarbenjamin I'm not being able to generate `html docs`. I have been trying to generate them for sometime, but everytime it fails to generate them. I'm using wsl2 on windows 10.

oscarbenjamin commented:
How exactly does it fail?

sidhu1012 commented:
1) imagemagick docbook2x graphviz
imagemagick: command not found

2) make html
rm -rf _build/logo
mkdir -p _build/logo
python ./generate_logos.py -d
make: python: Command not found
make: *** [Makefile:144: logo] Error 127

3) firefox _build/html/index.html
Unable to init server: Broadway display type not supported: yourmachine.yourdomain.com:0.0
Error: cannot open display: yourmachine.yourdomain.com:0.0

sidhu1012 commented:
@oscarbenjamin While working , I just found that `_eval_evalf` shouldn't be removed completely because it can be pretty useful in cases where we don't need `evalf()` but need mpmath functions , since they don't bother about `options` parameter.

sidhu1012 committed:
Update codegen

sidhu1012 committed:
Merge branch 'master' into eval

sidhu1012 committed:
Fix spacing

sidhu1012 committed:
Update sets

sidhu1012 committed:
Update docstring

sidhu1012 committed:
Update functions

sidhu1012 committed:
Revert "Update functions"

This reverts commit 9b340029fd377cef1f0d12a6b84b27b5c750f533.

sidhu1012 committed:
Update physics

sidhu1012 committed:
Merge branch 'master' into eval

sidhu1012 committed:
Fix code quality

czgdp1807 commented:
The code quality tests on this PR are failing.

sidhu1012 commented:
> The code quality tests on this PR are failing.

Yes I'll fix them soon, a bit busy these days.

sidhu1012 commented:
Feel free to take it over as I may not get much to complete it in upcoming months.
